draft-ietf-quic-tls.md

```diff
@@ -565,11 +565,19 @@ older than 1.3 is negotiated.
 
 ## ClientHello Size {#clienthello-size}
 
-QUIC requires that the first Initial packet from a client contain an entire
-cryptographic handshake message, which for TLS is the ClientHello.  Though a
-packet larger than 1200 bytes might be supported by the path, a client improves
-the likelihood that a packet is accepted if it ensures that the first
-ClientHello message is small enough to stay within this limit.
+The first Initial packet from a client starts with its first cryptographic
+handshake message, which for TLS is the ClientHello.  Servers might need to
+parse the entire ClientHello (e.g., to access extensions such as Server Name
+Identification (SNI) or Application Layer Protocol Negotiation (ALPN)) in order
+to decide whether to accept the new incoming QUIC connection.  If the
+ClientHello spans multiple Initial packets, such servers would need to buffer
+the first received fragments, which could consume excessive resources if the
+client's address has not yet been validated.  To avoid this, servers MAY use
+the Retry feature (see Section 8.1 of {{QUIC-TRANSPORT}}) to only buffer
+partial ClientHellos from clients with a validated address.  Though a packet
+larger than 1200 bytes might be supported by the path, a client improves the
+likelihood that a packet is accepted if it ensures that the first ClientHello
```

Comment:
s/is accepted/isn't dropped while in transit/

Reply:
I think that acceptance and loss are two separate things and what we're talking about here is acceptance.  The predicate on the statement is that the path supports packets of that size.  This is speculating about a server's willingness to buffer the ClientHello bytes.

To that end, this isn't about the 1200 byte minimum MTU, it's about making the ClientHello *as acceptable as possible*.  Fragments are immediately less acceptable, and while smaller fragments are easier to hold, what matters is the size of the entire message, because that's the upper limit on what the server might have to buffer.  Non-fragmented ClientHello can be as large as the path will support, because the server doesn't have to save them.

I think that we can strike this last sentence and things will be clearer.

Reply:
Related to "as acceptable as possible," should we try to recommend a minimum bound on fragments? (I'm reminded of data dribble, even though that's a different problem entirely.)

Reply:
If this multiple packet initial thing lands, at least we should limit the packet count / time to arrival or provide guidance. Dribble (slow loris) can happen in fragments or just in many very small complete packets over a very long time.

Reply:
Given that servers are encouraged to send Retry packets to clients with multiple-packet ClientHellos, this change doesn't make dribble worse, an attacker can send a single-packet ClientHello then start its SlowLoris without this change.

Reply:
You don't need to commit as much memory after that point.